pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -207,6 +207,13 @@ public void flush() {
         }
     }
 
+    @Override
+    public void flushAndClean() {
+        flush();
+        lastCumulativeAck = (MessageIdImpl) MessageId.earliest;
+        pendingIndividualAcks.clear();
```

Comment:
Doesn't flush empty out `pendingIndividualAcks`

https://github.com/apache/pulsar/blob/5f157a2a82eff0202a9fc0920bad67f5aabcde87/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java#L170

If yes then I would not create a new function - I would just add `lastCumulativeAck = (MessageIdImpl) MessageId.earliest;` to the flush() function.

Reply:
Yes, flush will empty `pendingIndividualAcks` but that doesn't mean that when you call `flush()` (do the effective redelivery of the message) also wants to lost the track of `lastCumulativeAck`.

I added `flushAndClean()` because here we need to start over again, the same as when the constructor occurs.